File: cpp/learning_cpp/general.cpp
/* general
    - syntaxes
        - statements
            - a statement is a single line (or compound) that performs a single operation
                - i.e. calling a function, initializing a variable, declaring a loop, etc.
            - every statement must end with a semicolon ;
            - example:
                std::string myString = "Hello World"; // declaring a variable
                std::cout << myString;          // print statement
        - using namespace
            - using namespace is used to omit certain namespaces in C++
                - reduces the time and complexity when calling certain namespaces
            - usually declared at the beginning of the program, below the header declarations
            - syntax:
                using namespace <namespace>;
            - example:
                using namespace std;

                // inside a function or method
                cout << "yes";
                // without using namespace
                std::cout << "yes"

        - comments
            - comments in C++ is done by declaring a double frontslash //
            - comments are not compiled into code
            - syntax:
                // <comment>
            - example:
                // comment here
        - case-sensitive
            - C++ is case-sensitive when it comes to naming
            - true and True are not the same thing
        - numbers and text
            - text in C++ must be wrapped in double quotes " "
                - a single character (char) is usually wrapped in single quotes ' '
                - example:
                    char myString[] = "idkman" // string
                    char myChar = 'Z'          // char
            - numbers doesn't need to be inside quotations
                - doing so will convert the number into a string
                - example:
                    int myNum = 100    // int
                    int myFloat = 3.14 // floats
        - code blocks
            - blocks or groups of code are often encased in curly bracer { }
            - it is recommended to encase blocks of code that is connected to each other in code blocks to group the,m
            - in an if, else if, else statements, and even in iterative statements
                - without a code block, only the first line will get executed as part of the statement
                    - any succeeding lines will get ignored or get executed outside of the statements
                - example:
                    if (true):
                        std::cout << "yes" // gets executed
                        std::cout << "no"  // does not get executed

    - naming conventions
        - there are industry standard when naming certain elements in C++
        - variables and attributes
            - usually in camelCase
            - example:
                int myString;
                char myNum[];
        - functions and methods
            - usually in camelCase
            - example:
                myFunction();
                sumAll();
        - filenames and folders
            - usually in snake_case
            - example:
                my_program.cpp
                idkman.cpp
        - enums, classes, structs and unions
            - usually in PascalCase
            - example:
                enum MyEnum {...};
                class MyClass {...};
                struct MyStruct {...};
        - pointers
            - usually in camelCase with a prefix of p
            - example:
                int *pnumPtr;
                float *pfloatPtr;
        - constants (const) and enum values
            - since enum values are similar to const variables, both follow the SCREAMING_SNAKE_CASE
            - example:
                const MY_VAR;
                enum MyEnum {
                    IDKMAN,
                    LUMBAGO
                }

    - scope
        - by default, the compiler compiles code from the top to bottom, first line until the last line
        - calling a variable that is not defined until later in the code will throw an error
        - example:
            std::cout << x; -> Error
            int x = 10;    // defined later while called before declared
        - block scope
            - are statements inside a block of code
            - variables inside a block scope only exists inside that code block
            - trying to access a variable inside a block scope from the outside will throw an error
            - example:
                void myFunc() {
                    int x = 10;
                }
                std::cout << x;        -> Error
            - loop scope
                - like block scope, variables inside a loop (e.g. for loops) are only accessible inside of it
                    - trying to access it outside the code block will throw an error
                - example:
                    while(true) {
                        int i = 10;
                        break;
                    }
                    std::cout << i -> Error

    - namespaces
        - is a way to group related code together under a name
        - helps in avoiding naming conflicts due to code size or library imports
        - why use namespaces?
            - avoids naming conflicts
                - mainly in large projects
            - organizes related code
                - groups them logically
            - separates code from code in libraries
        
        - declaration and definition
            - uses the namespace keyword
            - syntax:
                namespace <namespaceName> {...}
            - example:
                namespace idkman {
                    int yes = 10;
                }

        - access and namespace calls
            - members of a namespace are called using the double colon :: syntax
            - syntax:
                <namespaceName>::<member>;
            - example:
                idkman::yes;
            - to avoid typing the name of the namespace
                - it is paired using the "using" keyword before the main() function
                - syntax:
                    using namespace <namespaceName>;
                - example:
                    using namespace idkman;
                    // inside main()
                    yes;
                - note that is required to be careful when omitting namespaces due to possible conflicts

    - escape sequence
        - \n
            - the most common escape sequence
            - it represents a new line
        - \t
            - it represents a horizontal tab rule
        - \\
            - represents a single backslash as text
        - \"
            - represents a double quote ias text
*/

#include <iostream>
// header of the program
// you can import libraries here

int main() {
// main() denotes the main program
// the entry point of the program
    std::cout << "Hello World" << std::endl;
    // std::std::cout means print out, see-out
    // similar to printf() in C
    // semicolons ; are required to end statments
    // like C, errors will occur if statements don't end with ";"

    return 0;
    // marks the end of the program
    // 0 returns that the program ended with no errors
}
// basic C++ program

File: rust/src/learning/general.rs
/* general
    - syntaxes
        - statements
            - a statement is a single line (or compound) that performs a single operation
                - i.e. calling a function, initializing a variable, declaring a loop, etc.
            - every statement must end with a semicolon ;
            - example:
                println!("yes"); // printing text
        - comments
            - comments in C++ is done by declaring a double frontslash //
            - comments are not compiled into code
            - syntax:
                // <comment>
            - example:
                // comment here
        - placeholders
            - placeholders are used to show variable values
                - when iside a string or text variable
            - similar to Python's placeholders
            - uses curly braces {} followed by the placeholder value
            - syntax:
                "{}", <value>;
            - example:
                println!("idkman {}", "lumbago"); // idkman lumbago
            - note that placeholders and their respective values must be in order
                - example:
                    print!("{} is {}", "yes", 10); // yes is 10
            
    - naming conventions
        - there are industry standard when naming certain elements in C
        - variables
            - usually in snake_case
            - example:
                let my_string;
                let my_num[];
        - functions and methods
            - usually in snake_case
            - example:
                my_function();
                sum_all();
        - filenames and folders
            - usually in snake_case
            - example:
                my_program.rs
                idkman.rs
        - macros
            - usually in snake_case with an exclamation marks as a suffix
            - example:
                println!
                vec!
        - constants and static variables
            - usually in SCREAMING_SNAKE_CASE
            - example:
                CONSTANT_VARIABLE
                STATIC_VARIABLE
        - enum variants
            - usually in UpperCamelCase
            - example:
                Results::Ok
        - types, traits and enums
            - usually in UpperCamelCase
            - example:
                enum MyEnum {...}
                String

    - scope
        - by default, the compiler compiles code from the top to bottom, first line until the last line
        - calling a variable that is not defined until later in the code will throw an error
        - example:
            print!(x); -> Error
            let x: i32 = 10;    // defined later while called before declared
        - block scope
            - are statements inside a block of code
            - variables inside a block scope only exists inside that code block
            - trying to access a variable inside a block scope from the outside will throw an error
            - example:
                fn my_func() {
                    let x: i32 = 10;
                }
                print!(x);        -> Error
            - loop scope
                - like block scope, variables inside a loop (e.g. for loops) are only accessible inside of it
                    - trying to access it outside the code block will throw an error
                - example:
                    while(true) {
                        let i: i32 = 10;
                        break;
                    }
                    print!(i); -> Error

        - block labeling
            - blocks of lines are able to be labeled
                - used in loops to have precise control of looping
            - uses a single quote '
            - syntax:
                '<label>: {
                    <statements>
                }
            - example:
                'block: {
                    if 10 > 12 {
                        print!("yes");
                    } else {
                        print!("no");
                    }
                }
            - it is generally used to label loops
                - example:
                    'idkman: loop {
                        loop {
                            break 'idkman; // breaks the outer loop
                        }
                    }

        - shadowing 
            - variables in the same scope are able to be overshadowed
                - this is done by redeclaring the variable
            - example:
                let x: i32 = 10;
                let x: i32 = 100; // overshadows the first x variable
            - variables are able to be overshadowed inside a deeper scope
                - note that the outside variable is unaffected
                - example:
                    let x: i32 = 2;
                    {
                        let x: i32 = 10;
                    }
                    print!("{}", x);     // 2
            - it is recommended to have different names for each variable instead of overshadowing them
                - this makes the code easier to read

        - lifetime
            - Rust prevents dangling pointers by introducing lifetimes
                - a reference must not outlive the data it is pointing to
            - is the time it takes for a variable to be dropped
                - commonly when the variable hits the closing brace }
                - ending its lifetime
            - a reference is able to be tagged (or named)
                - multiple references are able to be tagged with the same name
                - with this, if one of the reference with the same name gets freed (or dropped)
                    - then the references become invalid
                - syntax:
                    '<tag>
                - example:
                    'a
                - sample code:
                    fn lifetimes<'a>(x: &'a str, y: &'a str) -> &'a str -{
                        a
                    }
                    // the output of lifetimes() is tied to the lifetime tag 'a
                    // if one of the references with the same tag goes out of scope
                    // then the anything with the tag 'a get's invalidated
                - sidenote:
                    - 'static is a reserved name
                        - used for strings that lasts until the end of the program
                    - '_ is an anonymous placeholder
*/


fn main() {
// main() acts as the entry point of the program
// anything inside its curly braces {} runs code
    print!("Hello World\n")
    // print! or println! is a macro
    // it prints text to the terminal
    // every line of code must end in semicolons ;
    // the program throws an error if not present
}
// basic Rust program
